core/shared/src/main/scala/zio/Scope.scala

```diff
@@ -224,188 +257,125 @@ object Scope {
      * The finalizer returned from this method will remove the original
      * finalizer from the map and run it.
      */
-    def add(finalizer: Finalizer)(implicit trace: Trace): UIO[Finalizer]
+    def add(finalizer: Finalizer)(implicit trace: Trace): UIO[Finalizer] =
+      modify {
+        case Running(nextKey, fins) =>
+          (
+            Exit.succeed(release(nextKey, _)),
+            Running(next(nextKey), fins.updated(nextKey, finalizer))
+          )
+        case Exited(nextKey, exit) =>
+          (
+            ZIO.suspendSucceed(finalizer(exit)) *> ReleaseMap.noopFinalizer,
+            Exited(next(nextKey), exit)
```

Comment:
out of curiosity, why can't you just return the state unmodified here?

Reply:
Frankly, I'm not 100% sure why, and to be on the safe side I decided to keep the implementation as close to the original as possible